### PR TITLE
guide: Add details to Basics section of SymPy's Architecture

### PR DESCRIPTION
**Pull request by kychua**
Add details explaining what goes on behind the syntactic sugar for
expressions. Adapted from part of the discussion from mailing list
(https://groups.google.com/forum/#!topic/sympy/4Pt1V322RHw/discussion).

I've added some explanation based on the discussion, not sure if it is too detailed or should be placed somewhere else though. 

<!-- If this pull request fixes an issue please indicate which issue by typing "Fixes #NNNN" below.>
Part of #5626

### CODE REVIEW

File: doc/src/guide.rst [hunk 1]
```diff
@@ -122,16 +122,27 @@ class from SymPy.  For example ``Add(Symbol("a"), Symbol("b"))`` gives an
...
+Then, in the ``__add__``, ``__mul__`` methods defined for SymPy objects, new
+Add or Mul objects are constructed. Python types are also converted to
+SymPy types through the ``sympify()`` function. Thus, ``cos(x).__add__(1)``
+returns  ``Add(cos(x), Integer(1))``.
+
+Similarly, ``2/cos(x)`` is equal to ``cos(x).__rdiv__(2)`` is equal to
 ``Mul(Rational(2), Pow(cos(x), Rational(-1)))``.
 
+Note that ``2/cos(x)`` calls ``cos(x).__rdiv__(2)`` instead of
+``2.__div__cos(x)`` because ``2`` (type ``int``) does not know how to divide
```
Comment (bjodah) [inline doc/src/guide.rst]: Typo: ".__div__cos(x)"

File: doc/src/guide.rst [hunk 2]
```diff
@@ -122,16 +122,27 @@ class from SymPy.  For example ``Add(Symbol("a"), Symbol("b"))`` gives an
...
 ``Add(cos(x), Integer(1))``
 
-or
+This works becauses Python allows you to override the behavior of built-in
+operations ``+, *, -, /`` etc. by defining methods with names ``__add__``,
+``__mul__`` etc, so ``cos(x) + 1`` is reduced to ``cos(x).__add__(1)``.
 
-``2/cos(x)`` is equal to ``cos(x).__rdiv__(2)`` is equal to
+Then, in the ``__add__``, ``__mul__`` methods defined for SymPy objects, new
+Add or Mul objects are constructed. Python types are also converted to
```
Comment (bjodah) [inline doc/src/guide.rst]: Maybe add double back-ticks around Mul and Add, e.g. \`\`Mul\`\` so that they get another style in the rendered HTML.

**Comment by siefkenj** [comment]
I think the Travis build failed for an unrelated reason.  It needs to be restarted.

File: doc/src/guide.rst [hunk 3]
```diff
@@ -122,16 +122,29 @@ class from SymPy.  For example ``Add(Symbol("a"), Symbol("b"))`` gives an
...
+For easier use, there is a syntactic sugar for expressions.
 
-``cos(x) + 1`` is equal to ``cos(x).__add__(1)`` is equal to
+For example, ``cos(x) + 1`` is equal to ``cos(x).__add__(1)`` is equal to
 ``Add(cos(x), Integer(1))``
 
-or
+This works becauses Python allows you to override the behavior of built-in
+operations ``+, *, -, /`` etc. by defining methods with names ``__add__``,
+``__mul__`` etc, so ``cos(x) + 1`` is reduced to ``cos(x).__add__(1)``.
```
Comment (siefkenj) [inline doc/src/guide.rst]: `etc.,`

Comment (siefkenj) [inline doc/src/guide.rst]: so ``cos(x)+1`` is reduced to 

->

so ``cos(x) + 1`` is evaluated as

File: doc/src/guide.rst [hunk 4]
```diff
@@ -122,16 +122,29 @@ class from SymPy.  For example ``Add(Symbol("a"), Symbol("b"))`` gives an
...
+``Add`` or ``Mul`` objects are constructed. Python types are also converted to
+SymPy types through the ``sympify()`` function. Thus, ``cos(x).__add__(1)``
+returns  ``Add(cos(x), Integer(1))``.
+
+Similarly, ``2/cos(x)`` is equal to ``cos(x).__rdiv__(2)`` is equal to
 ``Mul(Rational(2), Pow(cos(x), Rational(-1)))``.
 
+Note that ``2/cos(x)`` calls ``cos(x).__rdiv__(2)`` instead of
+``(2).__div__(cos(x))`` because ``2`` (type ``int``) does not know how to
+divide ``cos(x)``.
```
Comment (siefkenj) [inline doc/src/guide.rst]: "divide" -> "be divided by"

**Comment by siefkenj** [comment]
Looks ready to merge to me.

**Comment by gxyd** [comment]
Look good to me. I am merging this. Thanks @kychua .

**Pull request merged**